pkg/collector/corechecks/system/winregistry/winregistry.go

```diff
@@ -0,0 +1,191 @@
+	ValueIfMissing util.Optional[float64] `yaml:"value_if_missing, omitempty"`
+}
+
+type registryKeyCfg struct {
+	Name    string
+	Metrics map[string]metricCfg
+}
+
+type checkCfg struct {
+	RegistryKeys map[string]registryKeyCfg `yaml:"registry_keys"`
+}
+
+// registryKey is the in-memory representation of the key to monitor
+// it's different from the regstryKeyCfg because we need to split the hive
+// from the keypath. It's easier to do this once during the Configure phase,
+// than every time the check runs.
+type registryKey struct {
+	name            string
+	hive            registry.Key
+	keyPath         string
+	originalKeyPath string // keep the original keypath around, for logging errors
+	metrics         map[string]metricCfg
+}
+
+// WindowsRegistryCheck contains the field for the WindowsRegistryCheck
+type WindowsRegistryCheck struct {
+	core.CheckBase
+	metrics.Gauge
+	registryKeys []registryKey
+}
+
+func (c *WindowsRegistryCheck) Configure(integrationConfigDigest uint64, data integration.Data, initConfig integration.Data, source string) error {
+	err := c.CommonConfigure(integrationConfigDigest, initConfig, data, source)
+	if err != nil {
+		return err
+	}
+
+	var conf checkCfg
+	if err := yaml.Unmarshal(data, &conf); err != nil {
+		return err
+	}
+
+	hiveMap := map[string]registry.Key{
+		"HKLM":               registry.LOCAL_MACHINE,
+		"HKEY_LOCAL_MACHINE": registry.LOCAL_MACHINE,
+		"HKU":                registry.USERS,
+		"HKEY_USERS":         registry.USERS,
+		"HKEY_CLASSES_ROOT":  registry.CLASSES_ROOT,
+		"HKCR":               registry.CLASSES_ROOT,
+	}
+
+	for regKey, regKeyConfig := range conf.RegistryKeys {
+		splitKeypath := strings.Split(regKey, "\\")
+		if len(splitKeypath) < 2 {
+			return log.Errorf("the key %s is too short to be a valid key", regKey)
+		}
+
+		if hive, found := hiveMap[splitKeypath[0]]; found {
+			c.registryKeys = append(c.registryKeys, registryKey{
+				name:            regKeyConfig.Name,
+				hive:            hive,
+				originalKeyPath: regKey,
+				keyPath:         strings.Join(splitKeypath[1:], "\\"),
+				metrics:         regKeyConfig.Metrics,
+			})
+		} else {
+			return log.Errorf("unknown hive %s", splitKeypath[0])
+		}
+	}
+
+	return nil
+}
+
+func (c *WindowsRegistryCheck) Run() error {
+	sender, err := c.GetSender()
+	if err != nil {
+		return err
+	}
+
+	for _, regKey := range c.registryKeys {
+		k, err := registry.OpenKey(regKey.hive, regKey.keyPath, registry.QUERY_VALUE)
+		if err != nil {
+			if errors.Is(err, fs.ErrPermission) {
+				// Treat access denied as errors
+				log.Errorf("access denied while accessing key %s: %s", regKey.originalKeyPath, err)
+			} else if errors.Is(err, registry.ErrNotExist) {
+				log.Warnf("key %s was not found: %s", regKey.originalKeyPath, err)
+			}
+		} else {
+			// if err == nil the key was opened, so we need to close it after we are done.
+			//goland:noinspection GoDeferInLoop
+			defer func() {
+				k.Close()
+			}()
```

Comment:
If `registryKeys` can have a big number of elements I think it would be better to avoid deferring in a loop...

Do you need to keep `k` opened until the end of the function ?
If not you could just create an auxiliary function with the body of the loop and call it with the key argument, that way you can defer without risk and this function will be clearer.